dialer/replayer/replayer.go

```diff
@@ -0,0 +1,226 @@
+package replayer
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"time"
+
+	"github.com/gocql/gocql"
+	"github.com/gocql/gocql/dialer"
+)
+
+func NewReplayDialer(dir string) *ReplayDialer {
+	return &ReplayDialer{
+		dir: dir,
+	}
+}
+
+type ReplayDialer struct {
+	dir string
+	net.Dialer
+}
+
+func (d *ReplayDialer) DialContext(ctx context.Context, network, addr string) (conn net.Conn, err error) {
+	sourcePort := gocql.ScyllaGetSourcePort(ctx)
+	return NewConnectionReplayer(path.Join(d.dir, fmt.Sprintf("%s-%d", addr, sourcePort)))
+}
+
+func NewConnectionReplayer(fname string) (net.Conn, error) {
+	frames, err := loadResponseFramesFromFiles(fname+"Reads", fname+"Writes")
+	if err != nil {
+		return nil, err
+	}
+	return &ConnectionReplayer{frames: frames, frameIdsToReplay: []int{}, streamIdsToReplay: []int{}, frameIdx: 0, frameResponsePosition: 0, gotRequest: make(chan struct{}, 1)}, nil
+}
+
+type ConnectionReplayer struct {
+	frames                []*FrameRecorded
+	frameIdsToReplay      []int
+	streamIdsToReplay     []int
+	frameIdx              int
+	frameResponsePosition int
+	gotRequest            chan struct{}
+	closed                bool
+}
+
+func (c *ConnectionReplayer) frameStreamID() int {
+	return c.streamIdsToReplay[c.frameIdx]
+}
+
+func (c *ConnectionReplayer) getPendingFrame() *FrameRecorded {
+	if c.frameIdx < 0 || c.frameIdx >= len(c.frameIdsToReplay) {
+		return nil
+	}
+	frameId := c.frameIdsToReplay[c.frameIdx]
+	if frameId < 0 || frameId >= len(c.frames) {
+		return nil
+	}
+	return c.frames[frameId]
+}
+
+func (c *ConnectionReplayer) pushStreamIDToReplay(b []byte, idx int) {
+	if b[0] > 0x02 {
+		c.streamIdsToReplay = append(c.streamIdsToReplay, int(b[2])<<8|int(b[3]))
+	} else {
+		c.streamIdsToReplay = append(c.streamIdsToReplay, int(b[2]))
+	}
+	c.frameIdsToReplay = append(c.frameIdsToReplay, idx)
+}
+
+func replaceFrameStreamID(b []byte, stream int) {
+	if b[0] > 0x02 {
+		b[2] = byte(stream >> 8)
+		b[3] = byte(stream)
+	} else {
+		b[2] = byte(stream)
+	}
+}
+
+func (c *ConnectionReplayer) Read(b []byte) (n int, err error) {
+	frame := c.getPendingFrame()
+	for frame == nil {
+		// fmt.Println("FRAME NIL")
+		// <-c.gotRequest
+		// if c.Closed() {
+		// 	return 0, io.EOF
+		// }
+		// frame = c.getPendingFrame()
+		return 0, nil
+	}
+	response := frame.Response[c.frameResponsePosition:]
+
+	if len(b) < len(response) {
+		copy(b, response[:len(b)])
+		c.frameResponsePosition = c.frameResponsePosition + len(b)
+		return len(b), err
+	}
+
+	copy(b, response)
+	if c.frameResponsePosition == 0 {
+		replaceFrameStreamID(b, c.frameStreamID())
+	}
+
+	c.frameIdx = c.frameIdx + 1
+	c.frameResponsePosition = 0
+	return len(response), err
+}
+
+func (c *ConnectionReplayer) Write(b []byte) (n int, err error) {
+	writeHash := dialer.GetFrameHash(b)
+	// defer func() {
```

Comment:
It is better to move this code to the end of `pushStreamIDToReplay`, and please make it non-blocking